backupccl: error message for restoring databases with AOST before GC TTL

annezhu98 opened the pull request:
Previously, if a target database was not found for a RESTORE, whether it's due to AOST being before GC TTL or attempting to restore a database that has never existed, the same error message gets printed.

Since the errors do not occur for the same reason, having different error messages would be more informative for users to understand the cause of error.

Resolves: #47105

Release note (bug fix): more accurate error message for restoring AOST before GC TTL

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66025)
<!-- Reviewable:end -->

adityamaru commented:
Generally LGTM but can we add a test that looks something like the repro steps in the issue. 
Can we also consider adding something similar for table-level backups, I believe if you swap out the database backup/restore in the above reproduction steps we would see a different error when go to restore AOST a table that doesn't exist in our backup.

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=66025) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=66025) before we can accept your contribution.<br/>**11** out of **12** committers have signed the CLA.<br/><br/>:white_check_mark: rickystewart<br/>:white_check_mark: jordanlewis<br/>:white_check_mark: mgartner<br/>:white_check_mark: DrewKimball<br/>:white_check_mark: nathanstilwell<br/>:white_check_mark: jbowens<br/>:white_check_mark: dhartunian<br/>:white_check_mark: annezhu98<br/>:white_check_mark: knz<br/>:white_check_mark: miretskiy<br/>:white_check_mark: ajwerner<br/>:x: craig[bot]<br/><hr/>**craig[bot]** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=66025) it.</sub>

azhu-crl committed:
backupccl: error message for restoring databases with AOST before GC TTL

Previously, if a target database was not found for a RESTORE, whether it's due to AOST being before GC TTL or attempting to restore a database that has never existed, the same error message gets printed.

Since the errors do not occur for the same reason, having different error messages would be more informative for users to understand the cause of error.

Resolves: #47105

Release note (bug fix): more accurate error message for restoring AOST before GC TTL

knz commented:
nit: before this merges, ensure the commits are squashed and the fixup prefix has been removed.

adityamaru commented on pkg/ccl/backupccl/backupresolver/targets.go:
```diff
@@ -258,11 +259,16 @@ func DescriptorsMatchingTargets(
-			return ret, errors.Errorf("unknown database %q", d)
+			// given a non-empty aost, it's possible that we're unable to find the database due to aost being set before gc ttl
+			if asOf.IsEmpty() {
+				return ret, errors.Errorf("unknown database %q", d)
```
nit: can we change this and below to say - "database ... does not exist"

annezhu98 replied:
done

adityamaru commented:
Yup as Raphael mentioned above, we should amend the commit title to remove the fixup prefix, and squash the two commits into one. I can help you out with this, but should be good to go after that 🎉

adityamaru commented on pkg/ccl/backupccl/backupresolver/targets.go:
```diff
@@ -258,11 +259,16 @@ func DescriptorsMatchingTargets(
 		dbID, ok := resolver.DbsByName[string(d)]
 		if !ok {
-			return ret, errors.Errorf("unknown database %q", d)
+			// given a non-empty aost, it's possible that we're unable to find the database due to aost being set before gc ttl
```
nit: I think we can remove this comment, the error message is explanatory

annezhu98 replied:
done

annezhu98 commented:
bors r=adityamaru

annezhu98 commented:
bors r=adityamaru
